src/game.cpp

```diff
@@ -8015,47 +8016,62 @@ bool game::vehicle_near()
     return false;
 }
 
-bool game::refill_vehicle_part(vehicle &veh, vehicle_part *part, bool test)
+bool game::refill_vehicle_part(vehicle &veh, vehicle_part *part_ptr, bool test)
 {
-    vpart_info part_info = vehicle_part_types[part->id];
-    if (!part_info.has_flag("FUEL_TANK")) {
+(void)veh;(void)part_ptr;(void)test;
+    return false;// TODO: remove function
+}
+
+bool game::pl_refill_vehicle(vehicle &veh, int part, bool test)
+{
+    (void)test; // TODO: remove unused variable
+    
+    if( !veh.part_info( part ).has_flag("FUEL_TANK") ) {
         return false;
     }
-    item *it = nullptr; // the container or the fuel item,
-    item *p_itm = nullptr; // always the actual fuel item
-    long min_charges = -1;
+    vpart_info part_info = veh.part_info( part );
     bool in_container = false;
 
     std::string ftype = part_info.fuel_type;
-    itype_id itid = default_ammo(ftype);
-    if (u.weapon.is_container() && !u.weapon.contents.empty() &&
-        u.weapon.contents[0].type->id == itid) {
-        it = &u.weapon;
-        p_itm = &u.weapon.contents[0];
-        min_charges = u.weapon.contents[0].charges;
-        in_container = true;
-    } else if (u.weapon.type->id == itid) {
-        it = &u.weapon;
+    indexed_invslice reduced_inv;
+    if( ftype.empty() ) {
+        reduced_inv = u.inv.slice_filter_by( []( const item &itm ) {
+                return !itm.contents.empty() &&
+                       itm.contents.front().made_of( LIQUID );
+            } );
+    } else {
+        // Specifically handling containers made of liquid
+        // in case someone gets a really bad idea and makes it into a json
+        reduced_inv = u.inv.slice_filter_by( [ ftype ]( const item &itm ) {
+                if( !itm.contents.empty() ) {
+                    return itm.contents.front().typeId() == ftype;
+                } else {
+                    return itm.typeId() == ftype;
+                }
+            } );
+    }
+
+    int item_pos = display_slice(reduced_inv, _("Select item to refill with") );
+    
+    item *it = &( u.i_at( item_pos ) ); // Item or container
+    if( it == nullptr || it->is_null() ) {
+        return false;
+    }
+    item *p_itm; // Item used as fuel
+    if( it->contents.empty() ) {
         p_itm = it;
-        min_charges = u.weapon.charges;
     } else {
-        it = &u.inv.item_or_container(itid);
-        if (!it->is_null()) {
-            if (it->type->id == itid) {
-                p_itm = it;
-            } else {
-                //ah, must be a container of the thing
-                p_itm = &(it->contents[0]);
-                in_container = true;
-            }
-            min_charges = p_itm->charges;
-        }
+        p_itm = &it->contents.front();
+        in_container = true;
     }
-    // Check for p_itm->type->id == itid is already done above
-    if( p_itm == nullptr || it->is_null()) {
+
+    if( p_itm == nullptr || p_itm->is_null() ) {
```

Comment:
Please explain through which code path `p_itm` can end up being nullptr or being a null item (because I don't see any such path). Or just remove the check here.